df_py/volume/reward_calculator.py

```diff
@@ -0,0 +1,438 @@
+
+    @freeze_attributes
+    @enforce_types
+    def _reward_array_to_dicts(self) -> Tuple[dict, dict]:
+        """
+        @return
+          rewardsperlp -- dict of [chainID][LP_addr] : OCEAN_reward_float
+          rewardsinfo -- dict of [chainID][nft_addr][LP_addr] : OCEAN_reward_float
+
+        @notes
+          In the return dicts, chainID is the chain of the nft, not the
+          chain where rewards go.
+        """
+        rewardsperlp: dict = {}
+        rewardsinfo: dict = {}
+
+        for i, LP_addr in enumerate(self.LP_addrs):
+            for j, (chainID, nft_addr) in enumerate(self.chain_nft_tups):
+                assert self.R[i, j] >= 0.0, self.R[i, j]
+                if self.R[i, j] == 0.0:
+                    continue
+
+                if chainID not in rewardsperlp:
+                    rewardsperlp[chainID] = {}
+                if LP_addr not in rewardsperlp[chainID]:
+                    rewardsperlp[chainID][LP_addr] = 0.0
+                rewardsperlp[chainID][LP_addr] += self.R[i, j]
+
+                if chainID not in rewardsinfo:
+                    rewardsinfo[chainID] = {}
+                if nft_addr not in rewardsinfo[chainID]:
+                    rewardsinfo[chainID][nft_addr] = {}
+                rewardsinfo[chainID][nft_addr][LP_addr] = self.R[i, j]
+
+        return rewardsperlp, rewardsinfo
+
+    @freeze_attributes
+    @enforce_types
+    def _get_chain_nft_tups(self) -> List[Tuple[int, str]]:
+        """
+        @return
+          chain_nft_tups -- list of (chainID, nft_addr), indexed by j
+        """
+        chainIDs = list(self.stakes.keys())
+        nft_addrs = self._get_nft_addrs()
+        chain_nft_tups = [
+            (chainID, nft_addr)  # all (chain, nft) tups with stake
+            for chainID in chainIDs
+            for nft_addr in nft_addrs
+            if nft_addr in self.stakes[chainID]
+        ]
+        return chain_nft_tups
+
+    @freeze_attributes
+    @enforce_types
+    def _get_nft_addrs(self) -> List[str]:
+        """
+        @return
+          nft_addrs -- list of unique nft addrs. Order is consistent.
+        """
+        nft_addrs = set()
+        for chainID in self.nftvols_USD:
+            for nft_addr in self.nftvols_USD[chainID]:
+                nft_addrs.add(nft_addr)
+
+        return sorted(nft_addrs)
+
+    @freeze_attributes
+    @enforce_types
+    def _get_lp_addrs(self) -> List[str]:
+        """
+        @return
+          LP_addrs -- list of unique LP addrs. Order is consistent.
+        """
+        LP_addrs = set()
+        for chainID in self.stakes:
+            for nft_addr in self.stakes[chainID]:
+                for LP_addr in self.stakes[chainID][nft_addr]:
+                    LP_addrs.add(LP_addr)
+
+        return sorted(LP_addrs)
+
+    @freeze_attributes
+    @enforce_types
+    def _get_predictoors(self) -> Dict[int, List[str]]:
+        """
+        @return
+          predictoors -- dict of (chainID, list of nft_addrs)
```

Comment:
First: the name of this method is misleading. "Predictoors" are the people making predictions, or the bots that they're running. This method isn't returning predictoors. It's returning the contracts for the prediction feeds. 

To fix, while being aligned with query.py naming, I recommend renaming this method to: `_get_predictoor_contracts()`.

----
Next: It has super-important filtering behavior that's not obvious from its description or implementation.

I recommend to add to the description, something like:

```python
"""
@notes 
  This will only return the prediction feeds that are owned by DEPLOYER_ADDRS, due to functionality of query_predictoor_contracts().
"""
```

----
And, I see that `queries.py::query_predictoor_contracts()` doesn't have a similar note in its description. It should. Please add one.

----

Finally, I couldn't find a unit test for this method. It should have one. It could be simple, by mocking the call to `queries.py::query_predictoor_contracts()`.  

----
One more thing: are there other new functions or methods without unit tests? If yes, let's write basic tests for them too.

Reply:
@trentmc the filtering does not only return feeds owned by deployer addrs. It only uses the keys of deployer addrs, which is the list of chain ids where predictoor contracts are deployed. It is useless and error-prone to query contracts on chains where they are not deployed. I just added this to clarify.

In regards to unit tests, I think only this new function requires testing, otherwise the behaviour is tested (previous tests were adapted).

Reply:
Um, I'm pretty sure it _does_ filter. (If it doesn't, it's a bug.)

From `queries.py::query_predictoor_contracts()`:
```python
        for contract in predictoor_contracts:
            owner = contract["token"]["nft"]["owner"]["id"]
            if chain_id != DEV_CHAINID and owner not in DEPLOYER_ADDRS[chain_id]:
                continue
```